caddytls/ctmonitor.go

```diff
@@ -0,0 +1,298 @@
+	fmt.Printf("ct_id FilePath: %v\n", fileName)
+	indexBytes, err := ioutil.ReadFile(fileName)
+	if os.IsNotExist(err) {
+		log.Println("getLatestIndex failed, could not find the file")
+		return 0, nil
+	}
+	if err != nil {
+		return 0, err
+	}
+	indexStr := strings.TrimSpace(string(indexBytes))
+	index, err := strconv.Atoi(indexStr)
+	if err != nil {
+		log.Println("Error converting file string to int.")
+		return 0, err
+	}
+	return index, nil
+}
+
+func loadConfig() (config CtConfig) {
+	configJSON, err := os.Open(ctConfigFile)
+	if err != nil {
+		log.Printf("loadConfig error: %v", err)
+	}
+	defer configJSON.Close()
+	//var config Config
+	err = json.NewDecoder(configJSON).Decode(&config)
+	if err != nil {
+		log.Printf("[NOTICE] jsonDecode failed, error: %v\nUsing default values", err)
+		config.IncludeSubdomains = false
+		config.IncludeWildCards = false
+	}
+	return
+}
+
+// lookUpNames queries the certSpotter service for each SAN that Caddy is hosting
+// It then adds them to a set and returns a map of each certificate  mapped to a string
+// that contains identifying information for the cert.
+func lookUpNames(caddyCertsSANs []string, config QueryConfig) (map[string]string, int) {
+	// retrievedCerts is the bytes of a certificate mapped to the
+	// ID, issuer name, and before/after values.
+	retrievedCerts := make(map[string]string)
+
+	// biggestID is the most recent certificate id returned from CertSpotter.
+
+	// timeToWait is the amount of time you need to wait before querying again.
+	var biggestID, timeToWait int
+
+	// retryAfter is the timeToWait value from the response headers.
+	var retryAfter string
+
+	// If the caddyCertsSANs is empty, it shouldn't run this at all.
+	for _, domainName := range caddyCertsSANs {
+		retryAfter = queryDomainName(domainName, config, &biggestID, retrievedCerts)
+	}
+	err := putLatestID(biggestID, filePath)
+	if err != nil {
+		//return fmt.Errorf("writing latest ID: %v", err)
+		log.Printf("[WARNING] writing latest ID (%v): %v", biggestID, err)
+	}
+	timeToWait, err = strconv.Atoi(retryAfter)
+	if err != nil {
+		log.Printf(err.Error())
+		log.Print("Error retrieving time to wait, waiting 1 hour\n")
+		timeToWait = 3600
+	}
+	// returning retrievedCerts because it is created here and timeToWait
+	// so that I wait enough time before trying again.
+	return retrievedCerts, timeToWait
+}
+
+// monitorCerts continuously monitors the certificates that Caddy serves,
+// it queries again after the specified time.
+func monitorCerts() {
+	var queryConfig QueryConfig
+	config := loadConfig()
+	for {
+		namesToLookUp, caddyCerts := getCaddyCerts()
+		if len(namesToLookUp) == 0 {
+			log.Print("Could not retrieve DNS names from Caddy Certificate\n" +
+				"Make sure that you are serving on port 80 & 443\n" +
+				"Terminating monitorCerts.")
+			break
+		}
+		startingIndex, err := getLatestIndex(filePath)
+		if err != nil {
+			log.Printf("Error %v while getting starting index, starting at 0", err)
```

Comment:
Here again, the code continues on error -- is that intended?

Reply:
It was intended, instead of completely halting the process, it is given a default value of 0.  Unless they are including expired certs, which the default is to not include expired certs, it should get them close to what we were originally going to use.